File: server/src/routes/applications.ts
import { Router, Request, Response, NextFunction } from 'express';
import { eq, and } from 'drizzle-orm';
import { v4 as uuid } from 'uuid';
import db from '../db/index.js';
import { applications, influencers, campaigns, collaborations, brands } from '../db/schema.js';
import { authenticate, authorize } from '../middleware/auth.js';
import { AppError } from '../middleware/error.js';

const router = Router();

// List applications (filtered by role)
router.get('/', authenticate, async (req: Request, res: Response, next: NextFunction) => {
  try {
    const { campaignId, status } = req.query;
    let results;

    if (req.user!.role === 'influencer') {
      const inf = await db.select().from(influencers).where(eq(influencers.userId, req.user!.sub)).limit(1);
      if (!inf.length) throw new AppError('Profile not found', 404);
      results = await db.select().from(applications).where(eq(applications.influencerId, inf[0].id));
    } else {
      // Brand: get apps for their campaigns
      const brand = await db.select().from(brands).where(eq(brands.userId, req.user!.sub)).limit(1);
      if (!brand.length) throw new AppError('Profile not found', 404);
      const brandCampaigns = await db.select().from(campaigns).where(eq(campaigns.brandId, brand[0].id));
      const campaignIds = brandCampaigns.map(c => c.id);
      results = await db.select().from(applications);
      results = results.filter(a => campaignIds.includes(a.campaignId));
    }

    if (campaignId) results = results.filter(a => a.campaignId === campaignId);
    if (status) results = results.filter(a => a.status === status);

    res.json({ success: true, data: results });
  } catch (err) { next(err); }
});

// Create application (influencer only)
router.post('/', authenticate, authorize('influencer'), async (req: Request, res: Response, next: NextFunction) => {
  try {
    const { campaignId, proposal, proposedRate } = req.body;
    const inf = await db.select().from(influencers).where(eq(influencers.userId, req.user!.sub)).limit(1);
    if (!inf.length) throw new AppError('Profile not found', 404);

    // Check for duplicate
    const existing = await db.select().from(applications)
      .where(and(eq(applications.campaignId, campaignId), eq(applications.influencerId, inf[0].id))).limit(1);
    if (existing.length) throw new AppError('Already applied to this campaign', 409);

    const campaign = await db.select().from(campaigns).where(eq(campaigns.id, campaignId)).limit(1);
    if (!campaign.length) throw new AppError('Campaign not found', 404);

    const id = uuid();
    await db.insert(applications).values({
      id, campaignId, campaignTitle: campaign[0].title,
      influencerId: inf[0].id, influencerName: inf[0].displayName,
      influencerFollowers: inf[0].totalFollowers,
      influencerEngagement: inf[0].engagementRate,
      influencerCategory: inf[0].category,
      status: 'pending', proposal, proposedRate,
    });

    const result = await db.select().from(applications).where(eq(applications.id, id)).limit(1);
    res.status(201).json({ success: true, data: result[0] });
  } catch (err) { next(err); }
});

// Update application status (brand: accept/reject)
router.patch('/:id/status', authenticate, authorize('brand'), async (req: Request, res: Response, next: NextFunction) => {
  try {
    const { status } = req.body;
    if (!['accepted', 'rejected'].includes(status)) throw new AppError('Invalid status', 400);

    await db.update(applications).set({ status }).where(eq(applications.id, req.params.id));

    // If accepted, create collaboration
    if (status === 'accepted') {
      const app = await db.select().from(applications).where(eq(applications.id, req.params.id)).limit(1);
      if (app.length) {
        const campaign = await db.select().from(campaigns).where(eq(campaigns.id, app[0].campaignId)).limit(1);
        const brand = await db.select().from(brands).where(eq(brands.userId, req.user!.sub)).limit(1);
        if (campaign.length && brand.length) {
          await db.insert(collaborations).values({
            id: uuid(), campaignId: app[0].campaignId, campaignTitle: campaign[0].title,
            brandId: brand[0].id, brandName: brand[0].companyName,
            influencerId: app[0].influencerId, influencerName: app[0].influencerName,
            status: 'in_progress', agreedRate: app[0].proposedRate,
            deliverables: 'As per campaign requirements',
            startDate: campaign[0].startDate, endDate: campaign[0].endDate,
          });
        }
      }
    }

    res.json({ success: true, message: `Application ${status}` });
  } catch (err) { next(err); }
});

// Withdraw application (influencer)
router.post('/:id/withdraw', authenticate, authorize('influencer'), async (req: Request, res: Response, next: NextFunction) => {
  try {
    await db.update(applications).set({ status: 'withdrawn' }).where(eq(applications.id, req.params.id));
    res.json({ success: true, message: 'Application withdrawn' });
  } catch (err) { next(err); }
});

export default router;

File: Backend/routes/applications.py
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity, get_jwt
from models import db
from models.application import Application
from models.influencer import Influencer
from models.campaign import Campaign
from models.brand import Brand
from models.collaboration import Collaboration

applications_bp = Blueprint('applications', __name__)


@applications_bp.route('', methods=['GET'])
@jwt_required()
def list_applications():
    user_id = get_jwt_identity()
    claims = get_jwt()
    status_filter = request.args.get('status')
    campaign_filter = request.args.get('campaignId')

    if claims.get('role') == 'influencer':
        influencer = Influencer.query.filter_by(user_id=user_id).first()
        if not influencer:
            return jsonify(success=True, data=[]), 200
        query = Application.query.filter_by(influencer_id=influencer.id)
    else:
        brand = Brand.query.filter_by(user_id=user_id).first()
        if not brand:
            return jsonify(success=True, data=[]), 200
        campaign_ids = [c.id for c in brand.campaigns.all()]
        query = Application.query.filter(Application.campaign_id.in_(campaign_ids))

    if status_filter:
        query = query.filter(Application.status == status_filter)
    if campaign_filter:
        query = query.filter(Application.campaign_id == campaign_filter)

    apps = query.order_by(Application.created_at.desc()).all()
    return jsonify(success=True, data=[a.to_dict() for a in apps]), 200


@applications_bp.route('', methods=['POST'])
@jwt_required()
def create_application():
    claims = get_jwt()
    if claims.get('role') != 'influencer':
        return jsonify(success=False, error='Influencer role required'), 403

    data = request.get_json()
    user_id = get_jwt_identity()
    influencer = Influencer.query.filter_by(user_id=user_id).first()
    if not influencer:
        return jsonify(success=False, error='Profile not found'), 404

    campaign_id = data.get('campaignId')
    campaign = Campaign.query.get(campaign_id)
    if not campaign:
        return jsonify(success=False, error='Campaign not found'), 404

    existing = Application.query.filter_by(campaign_id=campaign_id, influencer_id=influencer.id).first()
    if existing:
        return jsonify(success=False, error='Already applied to this campaign'), 409

    app = Application(
        campaign_id=campaign_id,
        campaign_title=campaign.title,
        influencer_id=influencer.id,
        influencer_name=influencer.display_name,
        influencer_followers=influencer.total_followers,
        influencer_engagement=influencer.engagement_rate,
        influencer_category=influencer.category,
        status='pending',
        proposal=data.get('proposal', ''),
        proposed_rate=data.get('proposedRate', influencer.price_per_post or 0),
    )
    db.session.add(app)
    db.session.commit()
    return jsonify(success=True, data=app.to_dict()), 201


@applications_bp.route('/<app_id>/status', methods=['PUT'])
@jwt_required()
def update_status(app_id):
    claims = get_jwt()
    if claims.get('role') != 'brand':
        return jsonify(success=False, error='Brand role required'), 403

    data = request.get_json()
    new_status = data.get('status')
    if new_status not in ('accepted', 'rejected'):
        return jsonify(success=False, error='Status must be accepted or rejected'), 400

    app = Application.query.get(app_id)
    if not app:
        return jsonify(success=False, error='Application not found'), 404

    app.status = new_status
    db.session.flush()

    if new_status == 'accepted':
        campaign = Campaign.query.get(app.campaign_id)
        user_id = get_jwt_identity()
        brand = Brand.query.filter_by(user_id=user_id).first()
        if campaign and brand:
            collab = Collaboration(
                campaign_id=app.campaign_id,
                campaign_title=campaign.title,
                brand_id=brand.id,
                brand_name=brand.company_name,
                influencer_id=app.influencer_id,
                influencer_name=app.influencer_name,
                status='in_progress',
                agreed_rate=app.proposed_rate,
                deliverables='As per campaign requirements',
                start_date=campaign.start_date,
                end_date=campaign.end_date,
            )
            db.session.add(collab)

    db.session.commit()
    return jsonify(success=True, message=f'Application {new_status}'), 200


@applications_bp.route('/<app_id>/withdraw', methods=['POST'])
@jwt_required()
def withdraw(app_id):
    claims = get_jwt()
    if claims.get('role') != 'influencer':
        return jsonify(success=False, error='Influencer role required'), 403

    app = Application.query.get(app_id)
    if not app:
        return jsonify(success=False, error='Application not found'), 404

    app.status = 'withdrawn'
    db.session.commit()
    return jsonify(success=True, message='Application withdrawn'), 200
